[MRG+1] Add classes_ parameter to hyperparameter CV classes

stephen-hoover opened the pull request:
#### Reference Issue
Closes #8290 .

#### What does this implement/fix? Explain your changes.
In ``BaseSearchCV`` (superclass of ``GridSearchCV`` and ``RandomizedSearchCV``), add a ``classes_`` parameter which surfaces the ``classes_`` parameter of the ``best_estimator_``. Other parts of the scikit-learn code (e.g. ``sklearn.model_selection.cross_val_predict``) as well as users expect this property to be present on fitted classifiers.

#### Any other comments?
When making this PR, I realized that there had already been a nearly identical PR, #7661 . That PR added the ``classes_`` property in the deprecated ``sklearn.grid_search`` module. This PR adds it to the ``sklearn.model_selection`` module. The only difference is that I've added a
```python
self._check_is_fitted("classes_")
```
call, with the intent of raising an error which is clearer than an `AttributeError` referencing `best_estimator_`.

The What's New file claims that this change has already happened, so there doesn't seem to be anything to change there.

Commit:
ENH Add classes_ parameter to hyperparameter CV classes

In ``BaseSearchCV`` (superclass of ``GridSearchCV`` and ``RandomizedSearchCV``), add a ``clases_`` parameter which surfaces the ``classes_`` parameter of the ``best_estimator_``. Other parts of the scikit-learn code (e.g. ``cross_val_predict``) as well as users expect this property to be present on fitted classifiers.

stephen-hoover commented:
I don't know why Travis claims to have failed. It looks like all of the tests succeeded.

jnothman commented:
Travis is broken atm: https://github.com/travis-ci/travis-ci/issues/7264

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -914,7 +915,7 @@ def test_cross_val_predict_sparse_prediction():
 
 
-def test_cross_val_predict_with_method():
+def run_cross_val_predict_with_method(est):
```
we often call helpers in tests `check_*`

jnothman commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -254,6 +255,15 @@ def test_grid_search_groups():
+    grid_search = GridSearchCV(clf, {'foo_param': [1, 2, 3]})
+
+    grid_search.fit(X, y)
+    assert_array_equal(grid_search.classes_, np.unique(y))
```
Please add a test to check that `hasattr(gscv, 'classes_')` is false given a regressor?

jnothman commented:
Whoops that we failed to port across #7661 to `model_selection`!! Could you please port across the test too?

jnothman commented:
Thanks!

Commit:
Modify tests to address code review

stephen-hoover commented:
Thanks, @jnothman ! I copied the test from #7661 and added a couple of other checks on the `classes_` attribute. I also renamed the test helper function in `test_validation.py` to `check_cross_val_predict_with_method`.

jnothman reviewed (commented):
LGTM, thanks. And thanks for catching our omission.

Commit:
Merge remote-tracking branch 'upstream/master' into classes-param-for-basesearchcv

jnothman commented:
Usually I'd wait for another reviewer, just to be sure I didn't miss anything.

Could you add a note in what's new to say this was a bug fix because we failed to implement the change properly for 0.18?

Commit:
DOC Update What's New

stephen-hoover commented:
Okay. The change in #7661 was for v0.19; it hasn't been in a release yet. The text in the existing What's New entry describes what this PR does, so I added the PR number rather than making a new entry.

lesteve commented on doc/whats_new.rst:
```diff
@@ -68,7 +68,8 @@ Enhancements
    - Added ``classes_`` attribute to :class:`model_selection.GridSearchCV`
      that matches the ``classes_`` attribute of ``best_estimator_``. :issue:`7661`
-     by :user:`Alyssa Batula <abatula>` and :user:`Dylan Werner-Meier <unautre>`.
+     and :issue:`8295` by :user:`Alyssa Batula <abatula>`,
```
Maybe add the class `grid_search.GridSearchCV`, `grid_search.RandomizedSearchCV`  and `model_selection.RandomizedSearchCV` to make this entry more exact.

stephen-hoover replied:
Done.

Commit:
DOC Fix What's New addition

lesteve commented:
LGTM, thanks a lot, merging!

stephen-hoover commented:
Thank you!
